From 2b2f3f9bfcce3c42e3964bdbb6b66a99d9e10776 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Oliver=20St=C3=A4bler?= <oliver.staebler@bytesatwork.ch>
Date: Mon, 13 Jan 2014 16:16:04 +0100
Subject: [PATCH] bytePANEL: Add support for Chefree LVDS display

Add support for the Chefree CH70ILFL-001 LVDS display and introduce new
config parameter for LVDS and inverse polarity.
---
 arch/arm/configs/am335x_bytepanel_defconfig  |    6 ++++--
 arch/arm/mach-omap2/board-am335x-bytepanel.c |   26 ++++++++++++++++----------
 drivers/video/Kconfig                        |   14 ++++++++++++++
 drivers/video/da8xx-fb.c                     |   14 ++++++++++++++
 4 files changed, 48 insertions(+), 12 deletions(-)

diff --git a/arch/arm/configs/am335x_bytepanel_defconfig b/arch/arm/configs/am335x_bytepanel_defconfig
index 263f572..f7b0c0f 100644
--- a/arch/arm/configs/am335x_bytepanel_defconfig
+++ b/arch/arm/configs/am335x_bytepanel_defconfig
@@ -1785,6 +1785,10 @@ CONFIG_FB_CFB_IMAGEBLIT=y
 # CONFIG_FB_UDL is not set
 CONFIG_FB_DA8XX=y
 CONFIG_FB_DA8XX_CONSISTENT_DMA_SIZE=5
+CONFIG_LCD_INVERSE_POL=y
+CONFIG_LCD_ET070080DH6=y
+# CONFIG_LCD_DATAIMAGE is not set
+# CONFIG_LCD_CHEFREE is not set
 # CONFIG_FB_VIRTUAL is not set
 # CONFIG_FB_METRONOME is not set
 # CONFIG_FB_BROADSHEET is not set
@@ -1801,8 +1805,6 @@ CONFIG_LCD_CLASS_DEVICE=y
 # CONFIG_LCD_S6E63M0 is not set
 # CONFIG_LCD_LD9040 is not set
 # CONFIG_LCD_AMS369FG06 is not set
-CONFIG_LCD_ET070080DH6=y
-# CONFIG_LCD_DATAIMAGE is not set
 CONFIG_BACKLIGHT_CLASS_DEVICE=y
 # CONFIG_BACKLIGHT_GENERIC is not set
 CONFIG_BACKLIGHT_PWM=y
diff --git a/arch/arm/mach-omap2/board-am335x-bytepanel.c b/arch/arm/mach-omap2/board-am335x-bytepanel.c
index b846159..e1c4652 100644
--- a/arch/arm/mach-omap2/board-am335x-bytepanel.c
+++ b/arch/arm/mach-omap2/board-am335x-bytepanel.c
@@ -110,6 +110,12 @@ static struct da8xx_lcdc_platform_data lcdc_pdata = {
 	.controller_data	= &lcd_cfg,
 	.type			= "DATA_IMAGE",
 };
+#elif defined(CONFIG_LCD_CHEFREE)
+static struct da8xx_lcdc_platform_data lcdc_pdata = {
+	.manu_name		= "Chefree",
+	.controller_data	= &lcd_cfg,
+	.type			= "CHEFREE",
+};
 #endif
 
 /* LCD backlight platform Data */
@@ -399,9 +405,9 @@ static int __init backlight_init(void)
 	gpio_direction_output(LCD_BACKLIGHT_ENABLE, 1);
 	gpio_export(LCD_BACKLIGHT_ENABLE,0);
 
-#if defined( CONFIG_LCD_ET070080DH6)
+#if defined(CONFIG_LCD_INVERSE_POL)
 	pwm_pdata.chan_attrib[0].inverse_pol = true;
-#elif defined(CONFIG_LCD_DATAIMAGE)
+#else
 	pwm_pdata.chan_attrib[0].inverse_pol = false;
 #endif
 	am33xx_register_ecap(0, &pwm_pdata);
@@ -449,20 +455,20 @@ static void lcdc_init(void)
 	gpio_direction_output(LCD_DATA_ENABLE, 0);
 	gpio_export(LCD_DATA_ENABLE,0);
 
-#if defined(CONFIG_LCD_ET070080DH6)
-	/* Disable LVDS */
-	gpio_direction_output(LCD_LVDS_NSHUTDOWN, 0);
-	gpio_export(LCD_LVDS_NSHUTDOWN,0);
-	/* Enable LCD Panel PWR */
-	gpio_direction_output(LCD_PANEL_PWR, 1);
-	gpio_export(LCD_PANEL_PWR,0);
-#elif defined(CONFIG_LCD_DATAIMAGE)
+#if defined(CONFIG_LCD_LVDS)
 	/* Enable LVDS */
 	gpio_direction_output(LCD_LVDS_NSHUTDOWN, 1);
 	gpio_export(LCD_LVDS_NSHUTDOWN,0);
 	/* Disable LCD Panel PWR */
 	gpio_direction_output(LCD_PANEL_PWR, 0);
 	gpio_export(LCD_PANEL_PWR,0);
+#else
+	/* Disable LVDS */
+	gpio_direction_output(LCD_LVDS_NSHUTDOWN, 0);
+	gpio_export(LCD_LVDS_NSHUTDOWN,0);
+	/* Enable LCD Panel PWR */
+	gpio_direction_output(LCD_PANEL_PWR, 1);
+	gpio_export(LCD_PANEL_PWR,0);
 #endif
 
 	printk(KERN_INFO "Display initialized successfully\n");
diff --git a/drivers/video/Kconfig b/drivers/video/Kconfig
index 4932235..aee4244 100644
--- a/drivers/video/Kconfig
+++ b/drivers/video/Kconfig
@@ -2250,20 +2250,34 @@ config FB_DA8XX_CONSISTENT_DMA_SIZE
 	  If unsure say 1.
 
 if FB_DA8XX
+config LCD_LVDS
+	bool
+
+config LCD_INVERSE_POL
+	bool
+
 choice
 	prompt "bytePANEL Display Type"
 	default LCD_ET070080DH6
 
 config LCD_ET070080DH6
 	bool "Emerging ET070080DH6 (pRGB)"
+	select LCD_INVERSE_POL
 	help
 	  If you have an Emerging ET070080DH6 LCD Panel, say Y to enable its driver.
 
 config LCD_DATAIMAGE
 	bool "Data Image 10.4\" (LVDS)"
+	select LCD_LVDS
 	help
 	  If you have an Data Image LCD Panel, say Y to enable its driver.
 
+config LCD_CHEFREE
+	bool "Chefree CH70ILFL-001 (LVDS)"
+	select LCD_LVDS
+	help
+	  If you have an Chefree LCD Panel, say Y to enable its driver.
+
 endchoice
 endif
 
diff --git a/drivers/video/da8xx-fb.c b/drivers/video/da8xx-fb.c
index 5c7aa1b4..01a07da 100644
--- a/drivers/video/da8xx-fb.c
+++ b/drivers/video/da8xx-fb.c
@@ -318,6 +318,20 @@ static struct da8xx_panel known_lcd_panels[] = {
 		.pxl_clk = 47185000,
 		.invert_pxl_clk = 0,
 	},
+	/* CheFree CH70ILFL-001 LVDS */
+	[6] = {
+		.name = "CHEFREE",
+		.width = 1024,
+		.height = 600,
+		.hfp = 48,
+		.hbp = 240,
+		.hsw = 32,
+		.vfp = 3,
+		.vbp = 23,
+		.vsw = 12,
+		.pxl_clk = 51200000,
+		.invert_pxl_clk = 0,
+	},
 };
 
 static inline bool is_raster_enabled(void)
-- 
1.7.10.4

